### ISSUE TITLE
Allow functions to be declared at file level

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#8787]
It should be possible to declare functions at file level. Such functions would not have direct access to a contract, but they could call all other such functions defined or imported in the current file.

Such functions would have internal calling semantics, but could take storage pointers to act on the state.

The pseudo-keyword `this` would be invalid.

This could be a nice and more intuitive replacement for internal library functions.

**Comment by chriseth** [comment]
Should we also allow modifiers to be declared at file level?

**Comment by chriseth** [comment]
During the summit, a concern was raised that if this is used to replace internal library functions, the `using` construct will not be available here, but we could just allow `using <funcitonName> for X` or `using module.* for X` to attach any (set of) free function to a datatype.

**Comment by chriseth** [comment]
We might also introduce another namespacing construct to better group functions.

**Comment by chriseth** [comment]
As a proposal for the namespace, we could either use
```
namespace NAME { ... }
```
or
```
module NAME { ... }
```

**Comment by chriseth** [comment]
The namespace / module idea could be extended to template areas: https://github.com/ethereum/solidity/issues/9063

**Comment by ekpyron** [comment]
- Since we can actually introduce namespaces already be it only with the ``import`` construct, I think having ``namespace`` to allow this at a file level as well is a very good idea. Extending that to templated namespaces makes sense, but would require some more thought.

- I'm not quite sure about the ``using`` construct wrt to namespaces. E.g. what does the following mean?
  ```
  namespace X { struct Y { uint256 x; } }
  using X for uint256;
  ```
  This doesn't really make sense - and if we only allow it for individual functions, that is again subject to the concern of this being a hassle compared to internal library functions (i.e. "bulk" ``using`` statements for several functions). Allowing ``using`` only for namespaces that *only* contain functions, maybe even only functions with the correct type as first argument, might solve this, but I'm not sure about it.
- Can namespaces be re-opened/extended? If not, this might mean that ``namespace`` is not the best name, but I'm also not entirely sure about ``module``. If they can be re-opened, this would raise issues for restricting ``using`` to namespaces with only functions in them.
- I'd say whatever we allow for functions we should also allow for modifiers (i.e. have both of them at file level or in a construct like ``namespace``).
- A conservative approach to naming/doing this would be ``internal library {}``... not sure about it - just saying...

**Comment by chriseth** [comment]
using: Currently, libraries also contain lots of other things, most notably structs, but I would be fine with restricting `using` to only namespaces with only functions. I would not disallow functions with different first parameter type because it gives you some flexibility about how to group functions, but am also not too strong on this.

I would not allow namespaces to be re-opened.

About modifiers: it is probably not very useful to have a modifier that does not have access to state variables, but we can still allow it.

I think we should move away from the word library, it has been used for too many different things.

**Comment by chriseth** [comment]
Discussion: Let's implement the most basic version of this and try to re-write some existing contracts / libraries to make use of it.

**Comment by chriseth** [comment]
If we allow `using A.* for B` where `A` is an imported module without adding explicit namespaces or named modules, one tiny problem that still needs to be solved is if we want to allow `using A for B` at file level - as a replacement for it being used inside a library.

I think allowing `using A for B` at file level could also simplify the language. One issue with `using` we have is that it is not clear how it behaves with inheritance. I think it is more justifiable that `import "abc"` does not "import" the effects of "using" - and certainly `import "abc" as X` does not import it.

**Comment by chriseth** [comment]
Implemented in #8834

**Comment by Nielsbishere** [comment]
~Are constants also possible at file level? If not, why not?~ it seems to work, but visibility can't be used.
